File: src/commonTest/kotlin/com/scoreo/application/IdGeneratorTest.kt
package com.scoreo.application

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class IdGeneratorTest {

    private val UUID_REGEX = Regex(
        "^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
    )

    @Test
    fun `generated id matches UUID v4 format`() {
        val id = IdGenerator.newId()
        assertTrue(UUID_REGEX.matches(id), "ID '$id' does not match UUID v4 format")
    }

    @Test
    fun `generated id has correct length`() {
        val id = IdGenerator.newId()
        assertEquals(36, id.length)
    }

    @Test
    fun `generated id has version 4 at position 14`() {
        val id = IdGenerator.newId()
        assertEquals('4', id[14])
    }

    @Test
    fun `generated id has correct variant bits at position 19`() {
        val id = IdGenerator.newId()
        assertTrue(id[19] in listOf('8', '9', 'a', 'b'))
    }

    @Test
    fun `generated ids are unique`() {
        val ids = (1..100).map { IdGenerator.newId() }.toSet()
        assertEquals(100, ids.size, "Expected 100 unique IDs")
    }

    @Test
    fun `generated id has hyphens at correct positions`() {
        val id = IdGenerator.newId()
        assertEquals('-', id[8])
        assertEquals('-', id[13])
        assertEquals('-', id[18])
        assertEquals('-', id[23])
    }
}

File: apps/scoreo/src/application/idGenerator.test.ts
import { describe, expect, it } from 'vitest'
import { newId } from './idGenerator'

const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/

describe('IdGenerator', () => {
  it('generated id matches UUID v4 format', () => {
    const id = newId()
    expect(UUID_REGEX.test(id)).toBe(true)
  })

  it('generated id has correct length', () => {
    expect(newId()).toHaveLength(36)
  })

  it('generated id has version 4 at position 14', () => {
    expect(newId()[14]).toBe('4')
  })

  it('generated id has correct variant bits at position 19', () => {
    expect(['8', '9', 'a', 'b']).toContain(newId()[19])
  })

  it('generated ids are unique', () => {
    const ids = new Set(Array.from({ length: 100 }, () => newId()))
    expect(ids.size).toBe(100)
  })

  it('generated id has hyphens at correct positions', () => {
    const id = newId()
    expect(id[8]).toBe('-')
    expect(id[13]).toBe('-')
    expect(id[18]).toBe('-')
    expect(id[23]).toBe('-')
  })
})
